docs/linters.md

```diff
@@ -0,0 +1,210 @@
+problems getting something to lint, you can submit an unfinished PR
+and ask the reviewer to help you work through the lint problem, or you
+can tap into other Zulip community resources like our Google mailing list.
+
+Also, bear in mind that 100% of the lint code is open source, so if you
+find limitations in either the Zulip home-grown stuff or our third party
+tools, feedback will be highly appreciated.  (To give one example, as of
+this writing, we only recently recently started using mypy, which is an
+amazing project to provide static checking for Python code.  While mypy
+has been very good for us, it is still a young project, so you will see
+"type: ignore" comments in various places of our code.  As mypy evolves
+and certain alpha/beta bugs get fixed, we will sweep our code to improve
+annotations.)
+
+Finally, one way to clean up your code is to thoroughly exercise it
+with tests.  The [Zulip test documentation](testing.html)
+describes our test system in detail.
+
+## Lint checks
+
+Most of our lint checks get performed by `./tools/lint-all`.  These include the
+following checks:
+
+- Check Python code with pyflakes.
+- Check JavaScript code with jslint.
+- Check Python code for custom Zulip rules.
+- Check non-Python code for custom Zulip rules.
+- Check puppet scripts with the puppet validator.
+- Check HTML templates for matching tags and indentations.
+- Check CSS for parsability.
+- Check JavaScript code for addClass calls.
+
+The remaining lint checks occur in `./tools/run-mypy`.  It is probably somewhat
+of an understatement to call "mypy" a "linter," as it performs static
+code analysis of Python type annotations throughout our Python codebase.
+We will flesh out our documentation of mypy later.  The rest of this
+document pertains to the checks that occur in `./tools/lint-all`.
+
+## lint-all
+
+Zulip has a script called `lint-all` that lives in our "tools" directory.
+It is the workhorse of our linting system, although in some cases it
+dispatches the heavy lifting to other components such as pyflakes,
+jslint, and other home grown tools.
+
+You can find the source code [here](https://github.com/zulip/zulip/blob/master/tools/lint-all).
+
+In order for our entire lint suite to run in a timely fashion, the `lint-all`
+script performs several lint checks in parallel by forking out subprocesses.  This mechanism
+is still evolving, but you can look at the method `run_parallel` to get the
+gist of how it works.
+
+### Special options
+
+You can use the `-h` option for `lint-all` to see its usage.  One particular
+flag to take note of is the `--modified` flag, which enables you to only run
+lint checks against files that are modified in your git repo.  Most of the
+"sub-linters" respect this flag, but some will continue to process all the files.
+Generally, a good workflow is to run with `--modified` when you are iterating on
+the code, and then run without that option right before commiting new code.
```

Comment:
`--verbose` is nice for figuring out which linter is being slow, might be worth briefly documenting.